frontend/exporter/src/id_table.rs

```diff
@@ -0,0 +1,375 @@
+#[serde(transparent)]
+pub struct Id {
+    id: u32,
+}
+
+/// A session providing fresh IDs for ID table.
+#[derive(Default, Debug)]
+pub struct Session {
+    next_id: Id,
+    table: Table,
+}
+
+impl Session {
+    pub fn table(&self) -> &Table {
+        &self.table
+    }
+}
+
+/// The different types of values one can store in an ID table.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub enum Value {
+    Ty(Arc<TyKind>),
+}
+
+/// A node is a bundle of an ID with a value.
+#[derive(Deserialize, Serialize, Debug, JsonSchema, PartialEq, Eq, PartialOrd, Ord)]
+#[serde(into = "serde_repr::NodeRepr<T>")]
+#[serde(try_from = "serde_repr::NodeRepr<T>")]
+pub struct Node<T: 'static + SupportedType> {
+    id: Id,
+    value: Arc<T>,
+}
+
+/// Hax relies on hashes being deterministic for predicates
+/// ids. Identifiers are not deterministic: we implement hash for
+/// `Node` manually, discarding the field `id`.
+impl<T: SupportedType + Hash> Hash for Node<T> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.value.as_ref().hash(state);
+    }
+}
+
+/// Manual implementation of `Clone` that doesn't require a `Clone`
+/// bound on `T`.
+impl<T: SupportedType> Clone for Node<T> {
+    fn clone(&self) -> Self {
+        Self {
+            id: self.id.clone(),
+            value: self.value.clone(),
+        }
+    }
+}
+
+/// A table is a map from IDs to `Value`s. When serialized, we
+/// represent a table as a *sorted* vector. Indeed, the values stored
+/// in the table might reference each other, without cycle, so the
+/// order matters.
+#[derive(Default, Debug, Clone, Deserialize, Serialize)]
+#[serde(into = "serde_repr::SortedIdValuePairs")]
+#[serde(from = "serde_repr::SortedIdValuePairs")]
+pub struct Table(HashMap<Id, Value>);
+
+impl Table {
+    pub(super) fn insert<T>(&mut self, cache_id: Id, value: Arc<T>)
+    where
+        T: SupportedType,
+    {
+        self.0.insert(cache_id, T::to_types(value));
+    }
+    pub(super) fn get<T>(&self, cache_id: &Id) -> Option<Option<Arc<T>>>
+    where
+        T: SupportedType,
+    {
+        self.0.get(cache_id).map(T::from_types)
+    }
+}
+impl SupportedType for TyKind {
+    fn to_types(value: Arc<Self>) -> Value {
+        Value::Ty(value)
+    }
+    fn from_types(t: &Value) -> Option<Arc<Self>> {
+        match t {
+            Value::Ty(value) => Some(value.clone()),
+        }
+    }
+}
+
+/// A value encodable in the enum `Value`
+pub trait SupportedType: std::fmt::Debug {
+    fn to_types(value: Arc<Self>) -> Value;
+    fn from_types(t: &Value) -> Option<Arc<Self>>;
+}
+
+impl Session {
+    fn fresh_id(&mut self) -> Id {
+        let id = self.next_id.id;
+        self.next_id.id += 1;
+        Id { id }
+    }
+}
+
+impl<T: Sync + Send + Clone + 'static + SupportedType> Node<T> {
+    pub fn new(value: T, session: &mut Session) -> Self {
+        let id = session.fresh_id();
+        let kind = Arc::new(value);
```

Comment:
why not use `Arc::as_ptr(&kind).addr()` as unique identifier instead of making and storing a fresh one?

Reply:
Ah, that's a pleasant idea!
But looks like that requires a nightly feature.
I could do `(kind.as_ptr() as *const _) as usize` :thinking: Arc pins in memory I guess, that should be OK :thinking:

Reply:
That feature got stabilized 2 days ago, we can safely use it :)

Reply:
(`.as_ptr()` shouldn't work, you have to call it with `Arc::as_ptr()`. And the `as` cast is redundant, that method returns a `*const T` already)

Reply:
hax itself is using nightly, but we have a few other projects that use `hax-types` as a dependency, and those project are not nightly, that's a bit convoluted ^^'
Nice, will do that! That remove a type, that's good.

Reply:
Arf, no, I can't do that as such: I need an order on IDs!

Reply:
I mean, you can order `usize`s, do you also need that order to be stable or something like that?

Reply:
Currently, if a node has an id `N`, I know that its value might hold nodes whose id are strictly below `N`

So when I deserialize, I need to make sure to deserialize first the "leaf" nodes in the graph of nodes, so that I can incrementally store those in the state table. Then, nodes with higher IDs will be serialized, and might refer to nodes with lower IDs.
Do you see the point?

Reply:
Ah I see :thinking: You could use https://docs.rs/indexmap/latest/indexmap/ to remember the order of insertion: you're guaranteed that a leaf will be inserted before its parent. And just serialize in the order of insertion.

Reply:
Ah, that's tempting indeed! Gonna give that a shot.

Reply:
ah yes, I was thinking also I could just use a Vec and use the indices as fresh IDs

Reply:
That's even simpler than the addresses

Reply:
I don't know actually. I kind of like the fact table is the same thing during serialization and deserialization.
I think the current ID and Session thing is fine, I agree that we could optimize that a bit more, but I'm not sure wether that'd be much more clear.
Let's stick with that design for now

Reply:
Sure, it works :+1: